controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +110,288 @@ func NewReconciler(
+		}
+		if createdRequest.Status.ImageName != "" {
+			image.ImageName = createdRequest.Status.ImageName
+		}
+		updatedImages = append(updatedImages, image)
+	}
+
+	sort.Slice(updatedImages, func(i, j int) bool {
+		return updatedImages[i].Source < updatedImages[j].Source
+	})
+	return updatedImages
+}
+
+func (r *Reconciler) createNewVMPublishRequests(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	createdRequestsMap map[string]vmopv1.VirtualMachinePublishRequest,
+	desiredRequestsSet sets.Set[string]) error {
+
+	createdRequestsSet := sets.KeySet(createdRequestsMap)
+	toBeCreatedRequestsSet := desiredRequestsSet.Difference(createdRequestsSet)
+	gvk, err := apiutil.GVKForObject(ctx.VMGroupPublishRequest, r.Scheme())
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+	for toBeCreatedRequest := range toBeCreatedRequestsSet {
+		request := &vmopv1.VirtualMachinePublishRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    ctx.VMGroupPublishRequest.Namespace,
+				GenerateName: ctx.VMGroupPublishRequest.Name + "-",
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(ctx.VMGroupPublishRequest, gvk),
+				},
+			},
+			Spec: vmopv1.VirtualMachinePublishRequestSpec{
+				Source: vmopv1.VirtualMachinePublishRequestSource{
+					Name:       toBeCreatedRequest,
+					Kind:       reflect.TypeOf(vmopv1.VirtualMachine{}).Name(),
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+			},
+		}
+		if ctx.VMGroupPublishRequest.Spec.Target != "" {
+			request.Spec.Target = vmopv1.VirtualMachinePublishRequestTarget{
+				Location: vmopv1.VirtualMachinePublishRequestTargetLocation{
+					Name:       ctx.VMGroupPublishRequest.Spec.Target,
+					Kind:       reflect.TypeOf(imgregv1a1.ContentLibrary{}).Name(),
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+			}
+		}
+		if ctx.VMGroupPublishRequest.Spec.TTLSecondsAfterFinished != nil {
+			request.Spec.TTLSecondsAfterFinished = ctx.VMGroupPublishRequest.Spec.TTLSecondsAfterFinished
+		}
+		if err := r.Create(ctx, request); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (r *Reconciler) isCompleted(vmGroupPublishReq *vmopv1.VirtualMachineGroupPublishRequest,
+	pendingRequestsSet, completedRequestsSet, desiredRequestsSet sets.Set[string]) bool {
+
+	if completedRequestsSet.Equal(desiredRequestsSet) {
+		conditions.MarkTrue(vmGroupPublishReq, vmopv1.VirtualMachineGroupPublishRequestConditionComplete)
+		vmGroupPublishReq.Status.CompletionTime = metav1.Now()
+		return true
+	}
+
+	missingRequestsSet := desiredRequestsSet.Difference(completedRequestsSet.Union(pendingRequestsSet))
+
+	var reasons, messages []string
+	if len(missingRequestsSet) > 0 {
+		reasons = append(reasons, vmopv1.VirtualMachineGroupPublishRequestConditionReasonMissing)
+		messages = append(messages, fmt.Sprintf("waiting vm publish requests for %s to be created", strings.Join(missingRequestsSet.UnsortedList(), ", ")))
```

Comment:
FYI by using an `UnsortedList()` here and below, most likely the message is going to change on every reconcile regardless if any actual state has changed causing needless churn.

I don't think listing all the missing/pending requests in a condition message is all that useful - the underlying publish request state is better reflected in the group's Status (which I think it is, at least to some extent)

Reply:
make sense, updated the logic to only have a high-level count in the condition message.